### PR TITLE
feat (api): workflow joins

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/workflow/dao.go [hunk 1]
```diff
@@ -173,3 +201,71 @@ func updateLastModified(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) error
...
+		return sdk.NewError(sdk.ErrWorkflowInvalid, err)
+	}
+
+	//Checks application are in the current project
+	apps := w.InvolvedApplications()
+	for _, appID := range apps {
+		var found bool
+		for _, a := range proj.Applications {
+			if appID == a.ID {
+				found = true
```
Comment (yesnault) [inline engine/api/workflow/dao.go]: break

File: engine/api/workflow/dao.go [hunk 2]
```diff
@@ -173,3 +201,71 @@ func updateLastModified(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) error
...
+		}
+	}
+
+	//Checks pipelines are in the current project
+	pips := w.InvolvedPipelines()
+	for _, pipID := range pips {
+		var found bool
+		for _, p := range proj.Pipelines {
+			if pipID == p.ID {
+				found = true
```
Comment (yesnault) [inline engine/api/workflow/dao.go]: break

File: engine/api/workflow/dao.go [hunk 3]
```diff
@@ -173,3 +201,71 @@ func updateLastModified(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) error
...
+		}
+	}
+
+	//Checks environments are in the current project
+	envs := w.InvolvedPipelines()
+	for _, envID := range envs {
+		var found bool
+		for _, e := range proj.Environments {
+			if envID == e.ID {
+				found = true
```
Comment (yesnault) [inline engine/api/workflow/dao.go]: too

File: engine/api/workflow/dao_join.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+func deleteJoin(db gorp.SqlExecutor, n *sdk.WorkflowNodeJoin) error {
+	//Delete join triggers
+	for _, t := range n.Triggers {
+		if err := deleteJoinTrigger(db, &t); err != nil {
+			return sdk.WrapError(err, "deleteJoin> Unable to delete join trigger %d", t.ID)
+		}
+	}
+
+	//Delete associations between join and source
+	query := "delete from workflow_node_join_source where workflow_node_id = $1 and workflow_node_join_id = 2"
```
Comment (yesnault) [inline engine/api/workflow/dao_join.go]: not $2?

File: engine/api/workflow/dao.go [hunk 5]
```diff
@@ -112,25 +124,41 @@ func Insert(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) error {
...
 	if _, err := db.Update(&dbw); err != nil {
 		return sdk.WrapError(err, "Update> Unable to update workflow")
 	}
 	if w.Root != nil {
-		return insertOrUpdateNode(db, w, w.Root, u)
+		return insertOrUpdateNode(db, w, w.Root, u, false)
+	}
+	for _, j := range w.Joins {
+		if err := insertOrUpdateJoin(db, w, &j, u); err != nil {
+			return sdk.WrapError(err, "Insert> Unable to insert update workflow(%d) join (%#v)", w.ID, j)
```
Comment (yesnault) [inline engine/api/workflow/dao.go]: wrapError -> Update> (not Insert>)

File: engine/api/workflow/dao.go [hunk 6]
```diff
@@ -112,25 +124,41 @@ func Insert(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) error {
...
+		}
+	}
+
 	return updateLastModified(db, w, u)
 }
 
 // Update updates a workflow
 func Update(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) error {
+	if err := IsValid(db, w, u); err != nil {
+		return err
```
Comment (yesnault) [inline engine/api/workflow/dao.go]: no wrap?

File: engine/api/workflow/dao_join.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+)
+
+func loadJoins(db gorp.SqlExecutor, w *sdk.Workflow, u *sdk.User) ([]sdk.WorkflowNodeJoin, error) {
+	joinIDs := []int64{}
+	_, err := db.Select(&joinIDs, "select id from workflow_node_join where workflow_id = $1", w.ID)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		return nil, sdk.WrapError(err, "loadJoin> Unable to load join IDs on workflow %d", w.ID)
```
Comment (yesnault) [inline engine/api/workflow/dao_join.go]: loadJoins>

File: engine/api/workflow/dao_join.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+	}
+
+	joins := []sdk.WorkflowNodeJoin{}
+	for _, id := range joinIDs {
+		j, errJ := loadJoin(db, w, id, u)
+		if errJ != nil {
+			return nil, sdk.WrapError(errJ, "loadJoin> Unable to load join %d on workflow %d", id, w.ID)
```
Comment (yesnault) [inline engine/api/workflow/dao_join.go]: idem

File: engine/api/workflow/dao.go (hunk 6 above)
Comment (fsamin) [inline engine/api/workflow/dao.go]: nop, to keep readable error message coming from IsValid func

File: engine/api/workflow/dao.go (hunk 2 above)
Comment (fsamin) [inline engine/api/workflow/dao.go]: done

File: engine/api/workflow/dao.go (hunk 3 above)
Comment (fsamin) [inline engine/api/workflow/dao.go]: done

**Pull request merged**